User prefs2

tauhead opened the pull request:
Compared to the previous pull-request, this has moved all defaults to a top-level key.

There is also a new `fetch_userscript_preferences` function which returns the prefs from localStorage, merged with any default values.

tauhead committed:
fix broken options to hide "bribe for ration" & "buy a round".

* in-game markup must have changed

tauhead committed:
switch "extended nav" over to use new common preferences library - no more user-edited variables

tauhead committed:
comment-out example code

tauhead committed:
fix wrong variable name

tauhead committed:
change Extended Nav localStorage key-name. Fix heading typo

tauhead committed:
switch linkify-item-names.user.js to use userscript-preferences.js

tauhead committed:
change get_player_storage_prefix() to do username+"_"
instead of "_"+username

The code was originally copied from taustation-tools-common.js for generating prefixes to be used for lots of different keys.
We want to generate a nice keyname for a single use.
Now we can call get_player_storage_prefix( "myscript_prefs" );
to get the friendly key name "myscript_prefs_username".

tauhead committed:
standardize localStorage keys. Also move linkify 'debug' setting to bottom

quasidart commented:
Looks good! My only comments would be the ones I added to [the prior MR](https://github.com/taustation-fan/userscripts/pull/36).

tauhead committed:
add class to H2, and fix incorrect docs

tauhead committed:
add 'help' icon support

tauhead committed:
add help message

tauhead committed:
add help messages

tauhead committed:
fix TemperMonkey errors - don't declare variables in 'case' block

tauhead committed:
remove docs until design stabilizes

tauhead committed:
remove `callback` feature until design is agreed

tauhead commented:
Following on from the previous pull request (sorry, this got messy!)...

I think the way to get rid of the separate `defaults` arg, and get the default values back into each setting's own definition - and at the same time simplify the calling userscript's code - would be to have a single function used to both retrieve stored prefs, and add to the `/preferences` UI.

Something like:
`config = userscript_preferences( definition_object );`
and for cleanliness I'd probably have a function build the `definition_object` in my own code.
e.g.
`config = userscript_preferences( fetch_prefs_spec() );`

So the `userscript_preferences` function would:

  * Only if the URL is `/preferences`, use the definition to add the prefs UI
  * Merge localStorage & the definition to return a config object

tauhead commented:
We can get rid of the extra `get_player_storage_prefix` function by just using different keys of the definition object.

If there's a `key` just use that.

If there's a `player_key`, then call `get_player_storage_prefix` ourselves.

This way, callers only need to know about the single `userscript_preferences` function.

tauhead committed:
change to a single function that return prefs, and also add UI when at /preferences URL

tauhead committed:
change to use new single `userscript_preferences` function

tauhead committed:
change to use new single `userscript_preferences` function

tauhead committed:
change 'info-circle' icon to 'question-circle'

tauhead committed:
big refactor

- don't call localStorage.getItem() twice under /preferences URL
- cleaner jQuery element generation
- put all UI within a <span> so addition to DOM is atomic (avoid race between userscripts)
- bugfix: was putting each dt/dt pair in a new dl

tauhead committed:
the #preferences is a dl - shouldn't be creating our own

tauhead committed:
has bugs merging configs. also don't ignore false values

tauhead committed:
'array_checkbox' renamed to 'boolean_array'. same style as in-game. autosaves

tauhead committed:
rename 'array_checkbox' to 'boolean_array'

tauhead committed:
'checkbox' renamed to 'boolean'. same style as in-game. autosaves

tauhead committed:
rename 'checkbox' to 'boolean'

tauhead committed:
rename 'checkbox' to 'boolean'

tauhead committed:
text fields now autosave

tauhead committed:
text field has spinner while writing, and a green check after saving

tauhead committed:
radio fields now autosave. remove 'save' button

tauhead committed:
remove debugging

tauhead committed:
add spinner/check-mark to radio prefs. remove vertical stacking from radios

tauhead committed:
handle defined but false values properly

tauhead committed:
remove identical function

tauhead committed:
add 'select' pref type

tauhead committed:
change all 'data-userscript-pref' to 'name'

tauhead committed:
clean up donetyping() function to satify jshint & for readability

tauhead committed:
apply proper use of attr()/prop(). cast strings to numbers to avoid equality ==
